### PR TITLE
PR: Select full floating point numbers by double-clicking them

### LINKED ISSUE #22207: Feature: Select full floating point numbers by double-clicking on them
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [N/A] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [N/A] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [N/A] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
Often, my scripts print floating point numbers (like e.g. `0.7059913205389438`) to the IPython console, and often I want to copy some of those numbers.
Currently, I cannot easily select them by double-clicking.

(This is e.g. in contrast to the console of RStudio.)

### What steps reproduce the problem?

1.  Make a script print the above number to the console.
2.  Try to select the whole number for copying by double-clicking it.

### What is the expected output? What do you see instead?
Expected: The whole number is selected.
Observed: Only the decimals `7059913205389438` are selected.

## Versions

* Spyder version: 5.5.4  (pip)
* Python version: 3.10.12 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.10
* Operating System: Linux-6.5.0-41-generic-x86_64-with-glibc2.35

### Dependencies

<details>

```
# Mandatory:
atomicwrites >=1.2.0             :  1.4.1 (OK)
chardet >=2.0.0                  :  5.2.0 (OK)
cloudpickle >=0.5.0              :  3.0.0 (OK)
cookiecutter >=1.6.0             :  2.6.0 (OK)
diff_match_patch >=20181111      :  20230430 (OK)
intervaltree >=3.0.2             :  3.1.0 (OK)
IPython >=8.13.0,<9.0.0,!=8.17.1 :  8.24.0 (OK)
jedi >=0.17.2,<0.20.0            :  0.19.1 (OK)
jellyfish >=0.7                  :  1.0.3 (OK)
jsonschema >=3.2.0               :  4.22.0 (OK)
keyring >=17.0.0                 :  25.2.1 (OK)
nbconvert >=4.0                  :  7.16.4 (OK)
numpydoc >=0.6.0                 :  1.7.0 (OK)
parso >=0.7.0,<0.9.0             :  0.8.4 (OK)
pexpect >=4.4.0                  :  4.9.0 (OK)
pickleshare >=0.4                :  0.7.5 (OK)
psutil >=5.3                     :  5.9.8 (OK)
pygments >=2.0                   :  2.18.0 (OK)
pylint >=3.1,<4                  :  3.2.0 (OK)
pylint_venv >=3.0.2              :  3.0.3 (OK)
pyls_spyder >=0.4.0              :  0.4.0 (OK)
pylsp >=1.11.0,<1.12.0           :  1.11.0 (OK)
pylsp_black >=2.0.0,<3.0.0       :  2.0.0 (OK)
qdarkstyle >=3.2.0,<3.3.0        :  3.2.3 (OK)
qstylizer >=0.2.2                :  0.2.3 (OK)
qtawesome >=1.3.1,<1.4.0         :  1.3.1 (OK)
qtconsole >=5.5.1,<5.6.0         :  5.5.2 (OK)
qtpy >=2.1.0                     :  2.4.1 (OK)
rtree >=0.9.7                    :  1.2.0 (OK)
setuptools >=49.6.0              :  59.6.0 (OK)
sphinx >=0.6.6                   :  7.3.7 (OK)
spyder_kernels >=2.5.1,<2.6.0    :  2.5.1 (OK)
textdistance >=4.2.0             :  4.6.2 (OK)
three_merge >=0.1.1              :  0.1.1 (OK)
watchdog >=0.10.3                :  4.0.0 (OK)
xdg >=0.26                       :  0.28 (OK)
zmq >=24.0.0                     :  26.0.3 (OK)

# Optional:
cython >=0.21                    :  None (NOK)
matplotlib >=3.0.0               :  3.8.4 (OK)
numpy >=1.7                      :  1.26.4 (OK)
pandas >=1.1.1                   :  2.2.2 (OK)
scipy >=0.17.0                   :  1.13.0 (OK)
sympy >=0.7.3                    :  None (NOK)
```
</details>

### PR DESCRIPTION
**Pull request by athompson673**
Improve double click to select functionality from default to allow selecting floating point numbers, etc by parsing the line with the tokenize stdlib module and selecting entire NUMBER tokens.

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #22207

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: athompson673

### CODE REVIEW

File: spyder/widgets/mixins.py [hunk 1]
```diff
@@ -1499,6 +1502,27 @@ def _enter_array(self, inline):
                     self.sig_text_was_inserted.emit()
                 cursor.endEditBlock()
 
+    # ---- Qt methods
+    def mouseDoubleClickEvent(self, event):
+        """select NUMBER tokens to select numeric literals on double click"""
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:1507]:
```suggestion
        """Select NUMBER tokens to select numeric literals on double click."""
```

Minor improvements to this docstring.

File: spyder/widgets/mixins.py [hunk 2]
```diff
@@ -1499,6 +1502,27 @@ def _enter_array(self, inline):
...
+        cur = self.cursorForPosition(event.pos())
+        block = cur.block()
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:1509]:
```suggestion
        cursor = self.cursorForPosition(event.pos())
        block = cursor.block()
```

Rename `cur` to `cursor` for clarity.

File: spyder/widgets/mixins.py [hunk 3]
```diff
@@ -1499,6 +1502,27 @@ def _enter_array(self, inline):
...
+        text = block.text()
+        pos = block.position()
+        pos_in_block = cur.positionInBlock()
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:1512]:
```suggestion
        pos_in_block = cursor.positionInBlock()

```

For clarity.

File: spyder/widgets/mixins.py [hunk 4]
```diff
@@ -1499,6 +1502,27 @@ def _enter_array(self, inline):
...
+        for t_type, _, start, end, _ in generate_tokens(StringIO(text).read):
+            if t_type == NUMBER and start[1] <= pos_in_block <= end[1]:
+                cur.setPosition(pos + start[1])
+                cur.setPosition(pos + end[1], QTextCursor.MoveMode.KeepAnchor)
+                self.setTextCursor(cur)
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:1517]:
```suggestion
                cursor.setPosition(pos + start[1])
                cursor.setPosition(
                    pos + end[1], QTextCursor.MoveMode.KeepAnchor
                )
                self.setTextCursor(cursor)
```

`cur` to `cursor`.

File: spyder/widgets/mixins.py [hunk 5]
```diff
@@ -1499,6 +1502,27 @@ def _enter_array(self, inline):
...
+                return
+            elif start[1] > pos_in_block:
+                break
+        if isinstance(self, QPlainTextEdit):
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:1521]:
```suggestion

        if isinstance(self, QPlainTextEdit):
```

For clarity.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @athompson673 for opening your again PR against `master`!

I only left some minor style suggestions for you. In case you haven't used Github reviews before, you can submit all of them in a single commit by adding them to the batch, as described [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes).

**Comment by athompson673** [comment]
@ccordoba12 just ran into an issue with double clicking a docstring... I'll evaluate how to fix

image

**Comment by athompson673** [comment]
I figured after finding and solving that last issue, it would be a good idea to try to actually confirm there wouldn't be any more hidden problems. I wasn't sure how to add actual tests for automated testing, but I broke out some of the logic of the newly edited function, and ran a test against every possible double click position in all the .py files in a spyder 6.0.1 installation. No errors raised:

```python3
from pathlib import Path

from io import StringIO
from token import NUMBER
from tokenize import generate_tokens, TokenError

from time import perf_counter

def get_number_selection(line: str, pos: int) -> tuple[int, int]|tuple[None, None]:
    """
    Return start and end position of number literal at pos in line.
    If text at pos in line is not a number literal, return (None, None)
    """
    # Strip quotes to prevent tokenizer from trying to emit STRING tokens
    #   because we want to interpret numbers inside strings. This solves
    #   an EOF error trying to double click a line with opening or closing
    #   triple quotes as well.
    text = line.replace('"', ' ').replace("'", ' ')
    readline = StringIO(text).read
    
    try:
        for t_type, _, start, end, _ in generate_tokens(readline):
            if t_type == NUMBER and start[1] <= pos <= end[1]:
                return start[1], end[1]
            elif start[1] > pos:
                break

    except TokenError:
        # Ignore 'EOF in multi-line statement' from tokenize._tokenize
        # IndentationError should be impossible from tokenizing one line
        pass
    
    return None, None  # start and end of NUMBER not found

count = 0
start_time = perf_counter()
for p in Path(r'D:\python\py311\Lib\site-packages\spyder').glob("**/*.py"):
    with p.open(encoding="utf-8") as f:
        print(p)
        for line in f:
            for i in range(len(line)):
                start, end = get_number_selection(line, i)
                if start is None:
                    assert end is None
                elif end is None:
                    raise ValueError("1")
                else:
                    assert type(start) is int
                    assert type(end) is int
                    assert start < end
                    assert end <= len(line)
                count += 1
print("\n{count}, {perf_counter() - start_time}")

>>> %runfile D:/scripts/untitled0.py --wdir --post-mortem
5320059, 81.67973330011591
    
```

It also shows the function shouldn't slow down double click handling appreciably; about 15μs per call for me.

**Comment by ccordoba12** [comment]
> I wasn't sure how to add actual tests for automated testing, but I broke out some of the logic of the newly edited function, and ran a test against every possible double click position in all the .py files in a spyder 6.0.1 installation. No errors raised

That's great! Thanks for testing that @athompson673.

I'll give you a hand to implement a simple test for the new behavior you added, but just for the IPython console because it should be the same for the editor.

**Review by ccordoba12 — APPROVED** [review]
Thanks for your work on this @athompson673!! It's a really cool contribution!

**Comment by ccordoba12** [comment]
Note: This will be included in 6.1 (to be released in six months or so) because it's a new feature.

**Pull request merged**